text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,184 @@
+# Scenarios for Tracing semantic conventions for messaging
+
+This document aims to capture scenarios and a road map, both of which will
+serve as a basis for [stabilizing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable)
+the [existing semantic conventions for messaging](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md),
+which are currently in an [experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#experimental)
+state. The goal is to declare messaging semantic conventions stable before the
+end of 2021.
+
+## Motivation
+
+Many observability scenarios involve messaging systems. For Distributed Tracing
+to be useful across the entire scenario, having good observability for
+messaging operations is critical. To achieve this, OpenTelemetry must provide
+stable conventions and guidelines for instrumenting messaging systems.
+
+Bringing the existing experimental semantic conventions for messaging to a
+stable state is a crucial step for users and instrumentation authors, as it
+allows them to rely on [stability guarantees](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#not-defined-semantic-conventions-stability),
+and thus to ship and use stable instrumentation.
+
+## Roadmap
+
+| Description | Done By     |
+|-------------|-------------|
+| This OTEP, consisting of scenarios and a proposed roadmap, is approved and merged. | 09/30/2021 |
+| [Stability guarantees](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#not-defined-semantic-conventions-stability) for semantic conventions are approved and merged. This is not strictly related to semantic conventions for messaging but is a prerequisite for stabilizing any semantic conventions. | 09/30/2021 |
+| An OTEP proposing a set of attributes and conventions covering the scenarios in this document is approved and merged. | 10/29/2021 |
+| Proposed specification changes are verified by prototypes for the scenarios and examples below. | 11/15/2021 |
+| The [specification for messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) are updated according to the OTEP mentioned above and are declared [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable). | 11/30/2021 |
+
+## Terminology
+
+### Message
+
+A "message" is a transport wrapper for the transfer of information between two
```

Comment:
Not sure "transport" is appropriate as it's entirely possible for messages to be transferred between parties in process, which doesn't involve any transport.

For instance, Quarkus has the concept of "in memory" message channels that are equivalent to Kafka and AMQP but are entirely internal to an application

Reply:
In my understanding, the messaging semantic conventions aren't applicable to "in memory" message channels. I added this to the list of "Open questions", and I will also put it on the agenda for the next meeting.

Reply:
After discussion in today's WG meeting, I was asked to provide some more details on "in memory" to indicate how they align with messaging with a broker.

This behavior is based on my experience with Quarkus, so it may differ slightly from other frameworks.

An "in memory" channel has a unique name, within a single application, to define a source or destination of a message. With Quarkus, the only difference between an "in memory" channel and an external channel is the necessary configuration for the external broker, the code to consume/process/publish a message is identical.

With respect to the existing semantic conventions for messaging, there are attributes that would be equally applicable to "in memory" channels as much as ones interacting with an intermediary, such as `messaging.destination`. However, there are attributes that don't currently fit, or which might require an additional value, such as `messaging.destination`, which could possibly add `internal`.

If it's necessary for "in memory" channels to be out of scope for an initial release due to time constraints, that's ok provided it's taken into account by not including changes that would make it unworkable in the future. But I do think "in memory" should be in scope for these conventions as it can be just as important to observe the internal workings of a reactive application utilizing streams and event processing as a synchronous application with method calls.

I would equate "in memory" channel spans as the reactive equivalent to spans of synchronous method calls between services.

Reply:
> If it's necessary for "in memory" channels to be out of scope for an initial release due to time constraints, that's ok provided it's taken into account by not including changes that would make it unworkable in the future.

That sounds good to me. Do you think we could add Quarkus as specific product, alongside Kafka, RabbitMQ and others? I think it will be easy to support certain in-memory queues that behave like messaging systems, but I'm a bit hesitant to aim for general support for in-memory queues.

Reply:
Not sure if Quarkus really fits as a product from that perspective, maybe "Reactive Streams"? Might need some thought

Reply:
"In-memory" message channels are added to the Open Questions section, which we will go through in the workgroup once we're through the scenarios.

As long as Quarkus is covered by the scenarios, we should be able to support it with this semantic conventions. Probably we'd need to work on the wording: "reactive streams" sounds like a better category than "in-memory message channels". The latter, in my understanding, also incorporates channels as a programming language construct (like in Go), we don't want to apply the conventions to those.

Reply:
@kenfinnigan I tried to capture the results of our discussion in the document, adding this to the in-memory queues open question:
> While the semantic conventions might be used to instrument in-memory queues or channels which mimic the architecture of inter-application messaging systems, generally supporting in-memory queues or channels is considered out of scope.